File: src/main/java/no/sanchezrolfsen/framework/selenium/SeleniumUtils.java
package no.sanchezrolfsen.framework.selenium;

import io.cucumber.java.Scenario;
import lombok.experimental.UtilityClass;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.io.FileUtils;
import org.openqa.selenium.*;
import org.openqa.selenium.logging.LogEntries;
import org.openqa.selenium.logging.LogEntry;
import org.openqa.selenium.support.ui.ExpectedCondition;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;

import java.io.File;
import java.io.IOException;
import java.time.Duration;
import java.util.List;
import java.util.function.Consumer;

import static java.util.Objects.requireNonNull;
import static org.apache.commons.lang3.StringUtils.isBlank;
import static no.sanchezrolfsen.framework.selenium.TestUtils.doubleFromString;
import static no.sanchezrolfsen.framework.selenium.TestUtils.integerFromString;

@Slf4j
@UtilityClass
public class SeleniumUtils {

    public static final int MAX_ATTEMPTS = 4;
    public static final int WAIT_TIME = 1;

    public static Double getDoubleFromInput(WebElement el) {
        return doubleFromString(getInputText(el));
    }

    public static Integer getIntegerFromInput(WebElement el) {
        return integerFromString(getInputText(el));
    }

    public static WebElement waitForElementToBeClickable(WebElement element) {
        return waitForElementToBeClickable(element, Browser.globalWaitTimeout);
    }

    public static WebElement waitForElementToBeClickable(WebElement element, long timeout) {
        return Browser.pause(timeout)
                .ignoreAll(List.of(
                        NoSuchElementException.class, JavascriptException.class,
                        StaleElementReferenceException.class))
                .pollingEvery(Duration.ofMillis(100L))
                .until(ExpectedConditions.elementToBeClickable(element));
    }

    public static void clearInputWithSendKeys(WebElement webElement) {
        webElement.sendKeys(Keys.chord(Keys.CONTROL, "a"));
        webElement.sendKeys(Keys.BACK_SPACE);
    }

    public static WebElement waitFor(WebElement element) {
        return waitFor(element, Browser.globalWaitTimeout);
    }

    public static WebElement waitFor(WebElement element, long timeout) {
        return Browser.pause(timeout)
                .ignoreAll(List.of(StaleElementReferenceException.class, ElementNotVisibleException.class,
                        NoSuchElementException.class, JavascriptException.class))
                .pollingEvery(Duration.ofMillis(100L))
                .until(ExpectedConditions.visibilityOf(element));
    }

    public static WebElement waitFor(By locator) {
        return waitFor(locator, Browser.globalWaitTimeout);
    }

    public static WebElement waitFor(By locator, long timeout) {
        return Browser.pause(timeout)
                .ignoreAll(List.of(StaleElementReferenceException.class, ElementNotVisibleException.class,
                        NoSuchElementException.class, JavascriptException.class))
                .pollingEvery(Duration.ofMillis(100L))
                .until(ExpectedConditions.visibilityOfElementLocated(locator));
    }

    public static void waitForElementToHaveFocus(WebElement element) {
        waitForElementToHaveFocus(element, Browser.globalWaitTimeout);
    }

    public static void waitForElementToHaveFocus(final WebElement element, long timeout) {
        Browser.pause(timeout)
                .pollingEvery(Duration.ofMillis(100L))
                .until((ExpectedCondition<Boolean>) driver -> element.equals(requireNonNull(driver).switchTo().activeElement()));
    }

    public static void wait(int numberSeconds) {
        try {
            Thread.sleep(numberSeconds * 1000L);
        } catch (InterruptedException e) {
            e.printStackTrace();
            Thread.currentThread().interrupt();
        }
    }

    public static void retryingClick(WebElement element) {
        int attempts = 0;
        while (attempts < 5) {
            try {
                waitFor(element).click();
                return;
            } catch (Exception e) {
                log.info("Exception thrown, retrying click");
            }
            attempts++;
        }
    }

    public static void retryingSendKeys(WebElement element, String text) {
        int attempts = 0;
        while (attempts < 20) {
            try {
                element.sendKeys(text);
                break;
            } catch (Exception e) {
                log.info("Exception thrown, retrying sendKeys");
            }
            attempts++;
        }
    }

    public static void retryingSendKeys(By locator, String text) {
        int attempts = 0;
        while (attempts < 20) {
            try {
                Browser.vanillaDriver().findElement(locator).sendKeys(text);
                break;
            } catch (Exception e) {
                log.info("Exception thrown, retrying sendKeys");
            }
            attempts++;
        }
    }

    public static boolean stringListContainsString(List<String> elementList, String str) {
        return elementList.stream().anyMatch(s -> s.contains(str));
    }

    public static boolean elementContainsString(WebElement element, String str) {
        return element.getText().contains(str);
    }

    public static List<String> elementToStringList(List<WebElement> elements) {
        return elements.stream().map(WebElement::getText).toList();
    }

    public static boolean elementListContainsString(List<WebElement> elements, String str) {
        return stringListContainsString(elementToStringList(elements), str);
    }

    public static boolean acceptAlertIfPresent() {
        try {
            log.info("Waiting on JS-popup");
            Browser.pause(1).until(ExpectedConditions.alertIsPresent());
            closeAlert();
            return true;
        } catch (TimeoutException e) {
            log.debug(e.toString());
        }
        return false;
    }

    public static String getInputText(WebElement element) {
        return element.getAttribute("value");
    }

    public static void clearInputIfNotBlank(WebElement element) {
        if (!isInputBlank(element)) {
            element.clear();
        }
    }

    public static void clearInputIfNotEqual(WebElement el, String text) {
        if (!el.getText().equals(text))
            el.clear();
    }

    public static boolean isInputBlank(WebElement element) {
        return isBlank(getInputText(element));
    }

    public static void closeAlert() {
        Alert alert = Browser.vanillaDriver().switchTo().alert();
        alert.accept();
    }

    public static void safeSelect(final WebElement element, String valueToBeSelected) {
        new Select(waitFor(element)).selectByValue(valueToBeSelected);
    }

    public static boolean safeIsSelected(By by) {
        try {
            return Browser.vanillaDriver().findElement(by).isSelected();
        } catch (ElementNotVisibleException | NoSuchElementException | StaleElementReferenceException e) {
            log.debug(e.toString());
            return false;
        }
    }

    public static boolean safeIsVisible(By by) {
        try {
            return safeIsVisible(Browser.vanillaDriver().findElement(by));
        } catch (ElementNotVisibleException | StaleElementReferenceException | NotFoundException e) {
            return false;
        }
    }

    public static boolean safeIsVisible(WebElement element) {
        try {
            return !(element.getSize().getHeight() == 0 || element.getSize().getWidth() == 0);
        } catch (ElementNotVisibleException | StaleElementReferenceException | NotFoundException e) {
            log.debug("safeIsVisible: " + e);
            return false;
        }
    }

    public static boolean safeIsVisibleInside(WebElement parentElement, By by) {
        try {
            return !parentElement.findElement(by).getSize().equals(new Dimension(0, 0));
        } catch (ElementNotVisibleException | StaleElementReferenceException | NotFoundException e) {
            log.debug("safeIsVisible: " + e);
            return false;
        }
    }

    public static void embedScreenshot(Scenario scenario) {
        TakesScreenshot takesScreenshot = (TakesScreenshot) Browser.vanillaDriver();
        scenario.attach(takesScreenshot.getScreenshotAs(OutputType.BYTES), "image/png", "Screenshot");
        log.info("Screenshot of error included in the test-report");
    }

    public static void saveScreenshot(Scenario scenario, String filepath, String filename) {
        try {
            FileUtils.copyFile(((TakesScreenshot) Browser.vanillaDriver()).getScreenshotAs(OutputType.FILE),
                    new File(filepath, filename));
            log.info("Screenshot of error from scenario {} saved to {}/{}", scenario.getName(), filepath, filename);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void printBrowserLog() {
        LogEntries logs = Browser.vanillaDriver().manage().logs().get("browser");
        List<LogEntry> entries = logs.getAll();
        for (LogEntry entry : entries) {
            log.info(entry.toString());
        }
    }

    public static void scrollTo(WebElement webElement) {
        JavascriptExecutor javascriptExecutor = (JavascriptExecutor) Browser.vanillaDriver();
        javascriptExecutor.executeScript("arguments[0].scrollIntoView();", webElement);
    }

    public static void safeExecute(WebElement we, Consumer<WebElement> webElementConsumer) throws InterruptedException {
        boolean executed = waitLoop(we, webElementConsumer, WAIT_TIME, MAX_ATTEMPTS);
        if (!executed) {
            throw new InterruptedException("safeExecute failed");
        }
    }

    private static boolean waitLoop(WebElement we, Consumer<WebElement> webElementConsumer, int waitTime, int attempts) throws InterruptedException {
        int attempt = 0;
        boolean executed = false;
        while (attempt < attempts && !executed) {
            Browser.pause().until((ExpectedCondition<Boolean>) el -> SeleniumUtils.safeIsVisible(we));
            try {
                webElementConsumer.accept(we);
                executed = true;
            } catch (ElementNotVisibleException | StaleElementReferenceException  | NotFoundException e) {
                attempt += 1;
                SeleniumUtils.wait(waitTime);
            }
        }
        return executed;
    }

    public static void safeExecute(WebElement we, Consumer<WebElement> webElementConsumer, int wait, int attempts) throws InterruptedException {
        boolean executed = waitLoop(we, webElementConsumer, wait, attempts);
        if (!executed) {
            throw new InterruptedException("safeExecute failed");
        }
    }
}

File: src/main/kotlin/no/sanchezrolfsen/framework/selenium/SeleniumUtils.kt
package no.sanchezrolfsen.framework.selenium

import io.cucumber.java.Scenario
import org.apache.commons.io.FileUtils
import org.apache.commons.lang3.StringUtils.isBlank
import org.openqa.selenium.Alert
import org.openqa.selenium.By
import org.openqa.selenium.Dimension
import org.openqa.selenium.ElementNotInteractableException
import org.openqa.selenium.JavascriptException
import org.openqa.selenium.JavascriptExecutor
import org.openqa.selenium.NoSuchElementException
import org.openqa.selenium.NotFoundException
import org.openqa.selenium.OutputType
import org.openqa.selenium.StaleElementReferenceException
import org.openqa.selenium.TakesScreenshot
import org.openqa.selenium.TimeoutException
import org.openqa.selenium.WebElement
import org.openqa.selenium.support.ui.ExpectedCondition
import org.openqa.selenium.support.ui.ExpectedConditions
import org.openqa.selenium.support.ui.Select
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.time.Duration
import java.util.function.Consumer

object SeleniumUtils {

    private val log = LoggerFactory.getLogger(SeleniumUtils::class.java)

    const val MAX_ATTEMPTS = 4
    const val WAIT_TIME = 1

    @JvmStatic
    fun getDoubleFromInput(el: WebElement): Double? = TestUtils.doubleFromString(getInputText(el))

    @JvmStatic
    fun getIntegerFromInput(el: WebElement): Int? = TestUtils.integerFromString(getInputText(el))

    @JvmStatic
    @JvmOverloads
    fun waitForElementToBeClickable(element: WebElement, timeout: Long = Browser.globalWaitTimeout): WebElement =
        Browser.pause(timeout)
            .ignoreAll(
                listOf(
                    NoSuchElementException::class.java, JavascriptException::class.java,
                    StaleElementReferenceException::class.java
                )
            )
            .pollingEvery(Duration.ofMillis(100L))
            .until(ExpectedConditions.elementToBeClickable(element))

    @JvmStatic
    fun clearInputWithSendKeys(webElement: WebElement) {
        webElement.sendKeys(org.openqa.selenium.Keys.chord(org.openqa.selenium.Keys.CONTROL, "a"))
        webElement.sendKeys(org.openqa.selenium.Keys.BACK_SPACE)
    }

    @JvmStatic
    @JvmOverloads
    fun waitFor(element: WebElement, timeout: Long = Browser.globalWaitTimeout): WebElement =
        Browser.pause(timeout)
            .ignoreAll(
                listOf(
                    StaleElementReferenceException::class.java, ElementNotInteractableException::class.java,
                    NoSuchElementException::class.java, JavascriptException::class.java
                )
            )
            .pollingEvery(Duration.ofMillis(100L))
            .until(ExpectedConditions.visibilityOf(element))

    @JvmStatic
    @JvmOverloads
    fun waitFor(locator: By, timeout: Long = Browser.globalWaitTimeout): WebElement =
        Browser.pause(timeout)
            .ignoreAll(
                listOf(
                    StaleElementReferenceException::class.java, ElementNotInteractableException::class.java,
                    NoSuchElementException::class.java, JavascriptException::class.java
                )
            )
            .pollingEvery(Duration.ofMillis(100L))
            .until(ExpectedConditions.visibilityOfElementLocated(locator))

    @JvmStatic
    @JvmOverloads
    fun waitForElementToHaveFocus(element: WebElement, timeout: Long = Browser.globalWaitTimeout) {
        Browser.pause(timeout)
            .pollingEvery(Duration.ofMillis(100L))
            .until(ExpectedCondition { driver -> element == requireNotNull(driver).switchTo().activeElement() })
    }

    @JvmStatic
    fun wait(numberSeconds: Int) {
        try {
            Thread.sleep(numberSeconds * 1000L)
        } catch (e: InterruptedException) {
            e.printStackTrace()
            Thread.currentThread().interrupt()
        }
    }

    @JvmStatic
    fun retryingClick(element: WebElement) {
        var attempts = 0
        while (attempts < 5) {
            try {
                waitFor(element).click()
                return
            } catch (e: Exception) {
                log.info("Exception thrown, retrying click")
            }
            attempts++
        }
    }

    @JvmStatic
    fun retryingSendKeys(element: WebElement, text: String) {
        var attempts = 0
        while (attempts < 20) {
            try {
                element.sendKeys(text)
                break
            } catch (e: Exception) {
                log.info("Exception thrown, retrying sendKeys")
            }
            attempts++
        }
    }

    @JvmStatic
    fun retryingSendKeys(locator: By, text: String) {
        var attempts = 0
        while (attempts < 20) {
            try {
                Browser.vanillaDriver().findElement(locator).sendKeys(text)
                break
            } catch (e: Exception) {
                log.info("Exception thrown, retrying sendKeys")
            }
            attempts++
        }
    }

    @JvmStatic
    fun stringListContainsString(elementList: List<String>, str: String): Boolean =
        elementList.any { it.contains(str) }

    @JvmStatic
    fun elementContainsString(element: WebElement, str: String): Boolean = element.text.contains(str)

    @JvmStatic
    fun elementToStringList(elements: List<WebElement>): List<String> = elements.map { it.text }

    @JvmStatic
    fun elementListContainsString(elements: List<WebElement>, str: String): Boolean =
        stringListContainsString(elementToStringList(elements), str)

    @JvmStatic
    fun acceptAlertIfPresent(): Boolean {
        try {
            log.info("Waiting on JS-popup")
            Browser.pause(1).until(ExpectedConditions.alertIsPresent())
            closeAlert()
            return true
        } catch (e: TimeoutException) {
            log.debug(e.toString())
        }
        return false
    }

    @JvmStatic
    fun getInputText(element: WebElement): String? = element.getAttribute("value")

    @JvmStatic
    fun clearInputIfNotBlank(element: WebElement) {
        if (!isInputBlank(element)) {
            element.clear()
        }
    }

    @JvmStatic
    fun clearInputIfNotEqual(el: WebElement, text: String) {
        if (el.text != text) el.clear()
    }

    @JvmStatic
    fun isInputBlank(element: WebElement): Boolean = isBlank(getInputText(element))

    @JvmStatic
    fun closeAlert() {
        val alert: Alert = Browser.vanillaDriver().switchTo().alert()
        alert.accept()
    }

    @JvmStatic
    fun safeSelect(element: WebElement, valueToBeSelected: String) {
        Select(waitFor(element)).selectByValue(valueToBeSelected)
    }

    @JvmStatic
    fun safeIsSelected(by: By): Boolean {
        return try {
            Browser.vanillaDriver().findElement(by).isSelected
        } catch (e: ElementNotInteractableException) {
            log.debug(e.toString())
            false
        } catch (e: NoSuchElementException) {
            log.debug(e.toString())
            false
        } catch (e: StaleElementReferenceException) {
            log.debug(e.toString())
            false
        }
    }

    @JvmStatic
    fun safeIsVisible(by: By): Boolean {
        return try {
            safeIsVisible(Browser.vanillaDriver().findElement(by))
        } catch (e: ElementNotInteractableException) {
            false
        } catch (e: StaleElementReferenceException) {
            false
        } catch (e: NotFoundException) {
            false
        }
    }

    @JvmStatic
    fun safeIsVisible(element: WebElement): Boolean {
        return try {
            !(element.size.height == 0 || element.size.width == 0)
        } catch (e: ElementNotInteractableException) {
            log.debug("safeIsVisible: $e")
            false
        } catch (e: StaleElementReferenceException) {
            log.debug("safeIsVisible: $e")
            false
        } catch (e: NotFoundException) {
            log.debug("safeIsVisible: $e")
            false
        }
    }

    @JvmStatic
    fun safeIsVisibleInside(parentElement: WebElement, by: By): Boolean {
        return try {
            parentElement.findElement(by).size != Dimension(0, 0)
        } catch (e: ElementNotInteractableException) {
            log.debug("safeIsVisible: $e")
            false
        } catch (e: StaleElementReferenceException) {
            log.debug("safeIsVisible: $e")
            false
        } catch (e: NotFoundException) {
            log.debug("safeIsVisible: $e")
            false
        }
    }

    @JvmStatic
    fun embedScreenshot(scenario: Scenario) {
        val takesScreenshot = Browser.vanillaDriver() as TakesScreenshot
        scenario.attach(takesScreenshot.getScreenshotAs(OutputType.BYTES), "image/png", "Screenshot")
        log.info("Screenshot of error included in the test-report")
    }

    @JvmStatic
    fun saveScreenshot(scenario: Scenario, filepath: String, filename: String) {
        try {
            FileUtils.copyFile(
                (Browser.vanillaDriver() as TakesScreenshot).getScreenshotAs(OutputType.FILE),
                File(filepath, filename)
            )
            log.info("Screenshot of error from scenario {} saved to {}/{}", scenario.name, filepath, filename)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @JvmStatic
    fun printBrowserLog() {
        val logs = Browser.vanillaDriver().manage().logs().get("browser")
        val entries = logs.all
        for (entry in entries) {
            log.info(entry.toString())
        }
    }

    @JvmStatic
    fun scrollTo(webElement: WebElement) {
        val javascriptExecutor = Browser.vanillaDriver() as JavascriptExecutor
        javascriptExecutor.executeScript("arguments[0].scrollIntoView();", webElement)
    }

    @JvmStatic
    @JvmOverloads
    @Throws(InterruptedException::class)
    fun safeExecute(we: WebElement, webElementConsumer: Consumer<WebElement>, wait: Int = WAIT_TIME, attempts: Int = MAX_ATTEMPTS) {
        val executed = waitLoop(we, webElementConsumer, wait, attempts)
        if (!executed) {
            throw InterruptedException("safeExecute failed")
        }
    }

    private fun waitLoop(we: WebElement, webElementConsumer: Consumer<WebElement>, waitTime: Int, attempts: Int): Boolean {
        var attempt = 0
        var executed = false
        while (attempt < attempts && !executed) {
            Browser.pause().until(ExpectedCondition { safeIsVisible(we) })
            try {
                webElementConsumer.accept(we)
                executed = true
            } catch (e: ElementNotInteractableException) {
                attempt += 1
                wait(waitTime)
            } catch (e: StaleElementReferenceException) {
                attempt += 1
                wait(waitTime)
            } catch (e: NotFoundException) {
                attempt += 1
                wait(waitTime)
            }
        }
        return executed
    }
}
